### PR TITLE
Local Run/Debug now uses Globals section of SAM Templates

### PR DESCRIPTION
**Pull request by awschristou**
## Description

This change works with the Globals section of the SAM Template during local run/debug sessions.
For more information, see https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst

## Testing

* tests
* examined the generated "input template" when invoking node8/py37/dotnet21 both when the sam template had a Globals section and without it. The input template now copies over the Globals section when available.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/codelens/localLambdaRunner.ts [hunk 1]
```diff
@@ -358,6 +367,10 @@ export async function makeInputTemplate(params: {
         }
     }
 
+    if (params.globals) {
+        newTemplate.withGlobals(params.globals)
```
Comment (mpiroc) [inline src/shared/codelens/localLambdaRunner.ts]: I missed this in the first two reviews: This depends on the implementation detail that `withGlobals` returns `this`. To protect against implementation changes, suggest to use `newTemplate = newTemplate.withGlobals(...)`. Same comment for `withEnvironment` and `withMemorySize`.

File: src/shared/cloudformation/cloudformation.ts [hunk 2]
```diff
@@ -243,21 +249,34 @@ export namespace CloudFormation {
...
+        })
+    }
+
+    export async function getResourceFromTemplateResources(
+        params: {
+            templateResources?: TemplateResources,
+            handlerName: string,
+        }
+    ): Promise<Resource> {
+        const resources = params.templateResources || {}
```
Comment (bryceitoc9) [inline src/shared/cloudformation/cloudformation.ts]: I feel like this should be applied to line 254, and have templateResources be a mandatory parameter. Alternately, throw an error if templateResources is undefined.

File: src/test/shared/cloudformation/cloudformation.test.ts [hunk 3]
```diff
@@ -76,15 +86,31 @@ describe ('CloudFormation', () => {
...
+            const expectedTemplate = createBaseTemplate()
+            delete expectedTemplate.Globals
+
+            assert.deepStrictEqual(loadedTemplate, expectedTemplate)
+        })
+
+        it('can successfully load a file with parameters', async () => {
             const yamlStr: string =
-`Parameters:
+                `Parameters:
```
Comment (bryceitoc9) [inline src/test/shared/cloudformation/cloudformation.test.ts]: Was this (and other similar lines) fixed by the linter?

File: src/test/templates/sam/samTemplateGenerator.test.ts [hunk 4]
```diff
@@ -36,12 +36,16 @@ describe('SamTemplateGenerator', () => {
...
-        await new SamTemplateGenerator()
+    function makeMinimalTemplate(): SamTemplateGenerator {
+        return new SamTemplateGenerator()
             .withCodeUri(sampleCodeUriValue)
             .withFunctionHandler(sampleFunctionHandlerValue)
             .withRuntime(sampleRuntimeValue)
             .withResourceName(sampleResourceNameValue)
+    }
+
+    it('Produces a minimal template', async () => {
```
Comment (bryceitoc9) [inline src/test/templates/sam/samTemplateGenerator.test.ts]: `Produces a template with a MemorySize and Environment value`

File: src/shared/cloudformation/cloudformation.ts (hunk 2 above)
Comment (awschristou) [reply src/shared/cloudformation/cloudformation.ts]: template.yaml files aren't required to have a Resources section, so an undefined field is expected data.

File: src/test/shared/cloudformation/cloudformation.test.ts (hunk 3 above)
Comment (awschristou) [reply src/test/shared/cloudformation/cloudformation.test.ts]: Our repo doesn't have the formatter activated currently.
VS Code's autoformatting made adjustments to the file when I saved it.

File: src/test/templates/sam/samTemplateGenerator.test.ts (hunk 4 above)
Comment (awschristou) [reply src/test/templates/sam/samTemplateGenerator.test.ts]: I had a series of PRs for these different fields (MemorySize and Environment). Now that both are committed, I'll expand this test out to verify each separately.

File: src/shared/cloudformation/cloudformation.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/shared/cloudformation/cloudformation.ts]: Perhaps use `templateResources: TemplateResources = {}` in `params`?

**Pull request merged**